### DISCUSSION TITLE
Implementing SSO in Pimcore Asset Portal Engine

### DISCUSSION THREAD

**Discussion by Amarjeet664** [pimcore/pimcore#15776] [Q&A / Support]
Hi folks,

Has anyone implemented SSO on Pimcore Asset Portal Engine?

Already implemented the same on backend but wanted to know if we can implement the same at frontend on the login screen of Asset Portal Engine?

Is it the same process as adding in the backend?

**Comment by fashxp** [comment]
There is out-of-the box support for openID connect SSO. 
Besides that, there are Events https://pimcore.com/docs/platform/Portal_Engine/Development_Documentation/Customize_and_Extend_Behavior/Events#auth that should allow to integrate custom SSO solutions.

**Reply by Amarjeet664** [reply]
Thanks @fashxp 
We'll surely be looking into it.

**Reply by EliShklovsky** [reply]
Hi
we are struggling to configure the openID on the portal and we are facing problems.
The log error is:
invalid_client: XXXXX: The application 'XXXXXX' asked for scope 'groups' that doesn't exist on the resource 'XXXXX'. Contact the app vendor

Couple of question regarding the configuration:

1. The Redirect URL on the provider side should be:
https://<YOUR_PORTAL_DOMAIN>/auth/oidc/endpoint
or
https://<YOUR_PORTAL_DOMAIN>/auth/oidc/endpoint</PROVIDER_NAME>
(where PROVIDER_NAME is the name we configured on the portal OpenID connect Integration)

2. Should we first configure the Admin configuration (Settings->User->OpenIdConnect) in parallel to the portal OpenID Connect Integration? Is there any connection or dependency between the two configuration? 

3. Can we configure on the provider side settings dual redirect url ? for admin and for the portal?
Just for the info, we configured the admin and portal different domains like that:
admin domain: my.mydomain.com
portal domain: my-portal.mydoamin.com

4. Are there any special configurations when connecting to Microsoft Azure in terms of allowed scopes ? like Groups.Read.All etc.

Any answers will be appreciated,
Thanks
Eli

**Reply by fashxp** [reply]
Hi,
ad 1) Redirect url should be https://<YOUR_PORTAL_DOMAIN>/auth/oidc/endpoint
ad 2) There is no direct connection between these two
ad 3) Probably you need to setup two different apps on provider side
ad 4) I don't have experience with MS azure right now. But when the error message says scope `groups` does not exist, probably it is named differently in MS azure? Would be nice if you can share your experience then.

BR

**Reply by elements-team** [reply]
Hi,

We had a similar issue with openid through microsoft (openid connect) as you have to setup groups separately with this. as we didn't need group the solution for us was to disable the groups scope altogether (in the screenshot below we excluded groups from the scope). this way the user could login through openid

![image](https://github.com/user-attachments/assets/07aedf1a-bc51-4e09-9f87-2796bc57bbcd)

**Reply by EliShklovsky** [reply]
Hi
We are still struggling with no success with openid through MS Azure.
Is that configuration file was for the Admin or for the portal?
If we will configure the portal like that:

![image](https://github.com/user-attachments/assets/94f249db-cc2b-4440-8c73-7dd210430e24)

Would it act the same as with configuration file or should we explicitly create a conf file for the portal?
Thanks
Eli

**Reply by fashxp** [reply]
For portal engine, all SSO config takes place in the portal config in the document. there is no file for that...

**Reply by EliShklovsky** [reply]
@elements-team  - In addition to @fashxp comment, could you please explain how did you manage to implement this using config file? 
or did you do it on the Admin sso ?
Thanks
Eli

**Comment by EliShklovsky** [comment]
Eventually, after long time struggling with Microsoft Azure vs Pimcore's sso, where no group information received, and on each login the PortalUserGroup list of the PortalUser were automatically deleted,  we implemented a workaround using event listener of:

`event: pimcore.dataobject.preUpdate`
`event: pimcore.dataobject.postAdd`

where we set a default PortalUserGroup to the PortalUser

**Reply by kingjia90** [reply]
Could you please share the code of the workaround, i'll add a paragraph in the docs and hopefully would help anyone encountering similar situation with Azure

Thank you in advance

**Reply by EliShklovsky** [reply]
services.yaml

>     App\EventListener\AddDefaultGroupListener:
        arguments:
            $default_groupLogger: '@monolog.logger.default_group'
        tags:
            - { name: kernel.event_listener, event: pimcore.dataobject.preUpdate, method: onPostAdd }
            - { name: kernel.event_listener, event: pimcore.dataobject.postAdd, method: onPostAdd}



`<?php

namespace App\EventListener;

use Pimcore\Model\DataObject\PortalUser;
use Pimcore\Event\Model\DataObjectEvent;
use Pimcore\Model\DataObject\PortalUserGroup;
use Psr\Log\LoggerInterface;
use Symfony\Component\EventDispatcher\Attribute\AsEventListener;

class AddDefaultGroupListener
{
    private LoggerInterface $logger;

    public function __construct(LoggerInterface $default_groupLogger)
    {
        $this->logger = $default_groupLogger;
    }


    public function onPostAdd(DataObjectEvent $event): void
    {
        $object = $event->getObject();

        if ($object instanceof PortalUser) {
            $this->logger->info('PortalUser Created: ID ' . $object->getId());
            //Find the Group :
            $defaultGroup = PortalUserGroup::getByPath('/PortalEngine/Groups/group_portal_viewers');

            if ($defaultGroup) {
                $this->logger->info('Assigning Default Group: ' . $defaultGroup->getId());

                // Append Group Instead of Overwriting
                $existingGroups = $object->getGroups() ?: [];
                if (!in_array($defaultGroup, $existingGroups, true)) {
                    $existingGroups[] = $defaultGroup;
                    $object->setGroups($existingGroups);

                    //  Explicitly Save and Publish the Object
                    $object->save();
                    if (!$object->isPublished()) {
                        $this->logger->info('user added.');
                        $object->setPublished(true);
                        $object->save();
                    }

                    $this->logger->info('Group successfully added & user published.');
                } else {
                    $this->logger->info('User already has the group.');
                }
            } else {
                $this->logger->error('Default group not found!');
            }
        }
    }

}
`
